clang/lib/Driver/Driver.cpp

```diff
@@ -5134,6 +5121,49 @@ const ToolChain &Driver::getToolChain(const ArgList &Args,
   return *TC;
 }
 
+const ToolChain &Driver::getOffloadingDeviceToolChain(const ArgList &Args,
+                  const llvm::Triple &Target, const ToolChain &HostTC,
+                  const Action::OffloadKind &TargetDeviceOffloadKind) const {
+  // Use device / host triples as the key into the ToolChains map because the
+  // device ToolChain we create depends on both.
+  auto &TC = ToolChains[Target.str() + "/" + HostTC.getTriple().str()];
+  if (!TC) {
+    // Categorized by offload kind > arch rather than OS > arch like
+    // the normal getToolChain call, as it seems a reasonable way to categorize
+    // things.
+    switch (TargetDeviceOffloadKind) {
+      case Action::OFK_Cuda:
+        TC = llvm::make_unique<toolchains::CudaToolChain>(
+          *this, Target, HostTC, Args, TargetDeviceOffloadKind);
+        break;
+      case Action::OFK_HIP:
+        TC = llvm::make_unique<toolchains::HIPToolChain>(
+          *this, Target, HostTC, Args);
+        break;
+      case Action::OFK_OpenMP:
+        // omp + nvptx
+        TC = llvm::make_unique<toolchains::CudaToolChain>(
+          *this, Target, HostTC, Args, TargetDeviceOffloadKind);
+        break;
+      case Action::OFK_SYCL:
+        switch (Target.getArch()) {
```

Comment:
Not sure if this is the right place for target validation...

Reply:
I believe OpenMP checks for valid target combinations elsewhere and then rejects them if they're invalid, I can try and find the location again and add the equivalent for SYCL to the patch if you wish? Hopefully not misunderstanding what you mean in this case!

Reply:
I think CUDA also validates for target triple, but in some other place.

I think it would the right thing to do not diverge form the existing logic, unless we specifically trying to improve things.

I will appreciate if you can check if there is a better place to validate target triple by comparing with what CUDA/OpenMP compiler do.

Reply:
Yes, the check happens here: https://github.com/intel/llvm/blob/sycl/clang/lib/Frontend/CompilerInvocation.cpp#L2894 and this executes after a ToolChain has been chosen, but it's probably a more appropriate place to reject invalid targets. I am happy to make the change if you think it's the correct direction!

Reply:
This: https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/Driver.cpp#L2594 seems to be the CUDA equivalent. However, I think the OpenMP method is more inline with SYCL, especially as a somewhat equivalent check for SYCL (OpenMP also does the same check) appears to be done here: https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/Driver.cpp#L780